src/nnet3/nnet-discriminative-example.cc

```diff
@@ -0,0 +1,403 @@
+  for (int32 i = 0; i < num_examples; i++)
+    eg_inputs[i].io.swap((*input)[i].inputs);
+  // write to 'output->inputs'
+  eg_output.io.swap(output->inputs);
+
+  // Now deal with the discriminative-supervision 'outputs'.  There will
+  // normally be just one of these, with name "output", but we
+  // handle the more general case.
+  int32 num_output_names = (*input)[0].outputs.size();
+  output->outputs.resize(num_output_names);
+  for (int32 i = 0; i < num_output_names; i++) {
+    std::vector<const NnetDiscriminativeSupervision*> to_merge(num_examples);
+    for (int32 j = 0; j < num_examples; j++) {
+      KALDI_ASSERT((*input)[j].outputs.size() == num_output_names);
+      to_merge[j] = &((*input)[j].outputs[i]);
+    }
+    MergeSupervision(to_merge,
+                     &(output->outputs[i]));
+  }
+}
+
+void MergeSupervision(
+    const std::vector<const NnetDiscriminativeSupervision*> &inputs,
+    NnetDiscriminativeSupervision *output) {
+  int32 num_inputs = inputs.size(),
+      num_indexes = 0;
+  for (int32 n = 0; n < num_inputs; n++) {
+    KALDI_ASSERT(inputs[n]->name == inputs[0]->name);
+    num_indexes += inputs[n]->indexes.size();
+  }
+  output->name = inputs[0]->name;
+  std::vector<const discriminative::DiscriminativeSupervision*> input_supervision;
+  input_supervision.reserve(inputs.size());
+  for (int32 n = 0; n < num_inputs; n++)
+    input_supervision.push_back(&(inputs[n]->supervision));
+  std::vector<discriminative::DiscriminativeSupervision> output_supervision;
+  bool compactify = true;
+  discriminative::AppendSupervision(input_supervision,
+                         compactify,
+                         &output_supervision);
+  if (output_supervision.size() != 1)
+    KALDI_ERR << "Failed to merge discriminative examples-- inconsistent lengths "
+              << "or weights?";
+  output->supervision.Swap(&(output_supervision[0]));
+
+  output->indexes.clear();
+  output->indexes.reserve(num_indexes);
+  for (int32 n = 0; n < num_inputs; n++) {
+    const std::vector<Index> &src_indexes = inputs[n]->indexes;
+    int32 cur_size = output->indexes.size();
+    output->indexes.insert(output->indexes.end(),
+                           src_indexes.begin(), src_indexes.end());
+    std::vector<Index>::iterator iter = output->indexes.begin() + cur_size,
+        end = output->indexes.end();
+    // change the 'n' index to correspond to the index into 'input'.
+    // Each example gets a different 'n' value, starting from 0.
+    for (; iter != end; ++iter) {
+      KALDI_ASSERT(iter->n == 0 && "Merging already-merged discriminative egs");
+      iter->n = n;
+    }
+  }
+  KALDI_ASSERT(output->indexes.size() == num_indexes);
+  // OK, at this point the 'indexes' will be in the wrong order,
+  // because they should be first sorted by 't' and next by 'n'.
+  // 'sort' will fix this, due to the operator < on type Index.
+  std::sort(output->indexes.begin(), output->indexes.end());
```

Comment:
I was doing something really wrong when merging discriminative supervisions while training. What is the reason for this particular way of sorting in chain? This seems to not be the case in NnetIo structure. It seems more intuitive to have it the other way since the merged lattice and alignment are all stored in the order sorted by 'n' and then 't'.